Add port forwarding support to python-esiclient

larsks opened the pull request:
**This is a draft pull request, but I am looking for feedback**

Add the following commands:

- `openstack esi port forward create` -- create one or more port forwardings
- `openstack esi port forward delete` -- delete one or more port forwardings
- `openstack esi port forward purge` -- remove all port forwardings associated with a floating ip


## Create

The `create` and `delete` commands operate with forwarding specifications of the form `<internal ip>:<internal port>:<external ip>[:<external port>][/<protocol>`. For example, to create a port forward from floating ip address 100.100.100.100 port 22  to internal address 10.10.10.10 port 22:

```
openstack esi port forward create 10.10.10.10:22:100.100.100.100
```

This will:

- Find an existing port with the internal address 10.10.10.10, or
- If no port exists, (a) find an appropriate internal network for address 10.10.10.10 and create a new port
- Create the port forwarding from the floating ip to the internal ip

If the external port does not match the internal port, you can add an additional component to the specification, like this:

```
openstack esi port forward create 10.10.10.10:22:100.100.100.100:2222
```

This will create a port forwarding from 100.100.100.100 port 2222 to internal address 10.10.10.10 port 22.

You can create multiple forwardings in a single invocation:

```
openstack esi port forward create \
  10.10.10.10:22:100.100.100.100 \
  10.10.10.10:80:100.100.100.100 \
  10.10.10.10:443:100.100.100.100
```

## Delete

The syntax for the `delete` command is similar:

```
openstack esi port forward delete 10.10.10.10:22:100.100.100.100:2222
```

## Purge

The `purge` command removes all the port forwardings associated with a given floating ip address:

```
openstack esi port forward purge 100.100.100.100
```

tzumainn commented:
I kinda hate the number of colons in something like `10.10.10.10:443:100.100.100.100`, but I understand it's necessary if you want to be able to create a bunch of port forwardings in one ESI command. Is it necessary to do that though?

Will there be optional arguments for things like `protocol` or `description`?

If there are multiple internal networks that could match the given internal IP, would you error out?

I think you initially mentioned the idea of having a floating IP address automatically created for you; is that something you'd still want to do? Perhaps you could specify a network name instead of an IP, and have Neutron create the port for you?

larsks commented:
> I kinda hate the number of colons in something like `10.10.10.10:443:100.100.100.100`...

That's similar to the syntax we use for docker/podman port forwarding (`-p addr:external_port:internal_port`) or ssh port forwarding (`-L [local_address:]localhost_port:remote_host:remote_port`).

> Is it necessary to do that though?

I think it's very common to want to create multiple port forwards for baremetal hosts (e.g., the example from the description, in which we're forwarding ssh, http, and https).

> Will there be optional arguments for things like `protocol` or `description`?

Protocol can already be specified in a fashion similar to docker/podman:

```
10.10.10.10:67:100.100.100.100/udp
```

I haven't been able to figure out a good way of setting a description per port-forward; I guess you specify a single description that would be used for all of them.

> If there are multiple internal networks that could match the given internal IP, would you error out?

It does not currently error out. That would be a relatively easy change. You can specify an explicit internal network (and subnet) using `--internal-ip-network` and `--internal-ip-subnet` options.

> I think you initially mentioned the idea of having a floating IP address automatically created for you; is that something you'd still want to do? Perhaps you could specify a network name instead of an IP, and have Neutron create the port for you?

I ended up abandoning that because (a) it's not possible to create *specific* addresses, and (b) we have pretty strict quotas on floating ips and I wasn't sure if that would interact badly with a command that automatically allocates floating ips. I'm not dead set against it, though. See below for some thoughts.

---

If we assume that the most common case will be creating a set of port forwardings targeting the same external ip/internal ip pair, we could maybe do something like this instead:

```
openstack esi port forward create [--from <floating ip>|--create-floating-ip] <internal ip> [<outside port>:]inside_port[/<protocol>] [...]
```

Where the earlier example would become:

```
openstack esi port forward create --from 100.100.100.100 10.10.10.10 22 80 443
```

Or, to create a new floating ip:

```
openstack esi port forward create --create-floating-ip 10.10.10.10 22 80 443
```

The `purge` command would remain unmodified, and the `delete` command would mirror the `create` command:

```
openstack esi port forward delete --from 100.100.100.100 10.10.10.10 22 80 443
```

There would obviously be no `--create-floating-ip` option in this case, and `--from` would be a required option (and perhaps `--from *` would delete any matching port forwardings).

tzumainn commented:
I'm vaguely - but honestly, only vaguely - bothered by how different this is from typical OpenStack CLI syntax, in that the OpenStack CLI loves to use argument flags to make it clear what's being passed in. I'm not a huge fan of moving entirely towards docker port forwarding syntax, because there is still (currently) the expectation that the user uses other non-ESI OpenStack CLIs, and I do think it should feel "similar" - but I see that you are compromising on that in your examples.

I think it might feel more `OpenStack-y` if you did something like the following?

`openstack esi port forward create --port <protocol port designation 1>,<protocol port designation 2>,... <internal network designation> <external network designation>`

The comma separated list of protocol ports would either be `X` or `X:Y` to make it simple for the typical case where the internal/external port is the same, while allowing for the case where it's different.

The network designation could either be an IP address or a Neutron port name, or a Neutron network name; in the latter case, a port (or floating IP address) is created for you (and if it happens that a port and network - or IP address? - have the same name, then we throw an error).

We allow users to pass in `--protocol` if they want something other than tcp, and a `--description`. If they want different values for these, then... they just have to run the command multiple times. I could possibly see the protocol being wrapped up in the protocol port designation in the way that you describe above.

larsks commented:
> The internal network designation could either be an IP address or a Neutron port name, or a Neutron network name; in the latter case, a Neutron port is created for you (and if it happens that a port and network - or IP address? - have the same name, then we throw an error).

I think it only makes sense for the internal network designation to be either an ip address or an existing port uuid/name; I don't think it makes sense to create a forward to a random internal address.

> openstack esi port forward create --port <protocol port designation 1>,<protocol port designation 2>,... --from <external network designation> --to <internal network designation>

The thing that bothers me here is that both `--to` and `--from` are required. In the general scheme of things, this suggests they should be positional arguments rather than option arguments...

```
openstack esi port forward create [--port <portspec> [...]] <internal network designation> <external network designation>
```

...but that loses the contextual clues provided by the option names. Do you have a preference?

> What do you think?

I think that overall it sounds like a reasonable plan.

tzumainn commented:
> > The internal network designation could either be an IP address or a Neutron port name, or a Neutron network name; in the latter case, a Neutron port is created for you (and if it happens that a port and network - or IP address? - have the same name, then we throw an error).
> 
> I think it only makes sense for the internal network designation to be either an ip address or an existing port uuid/name; I don't think it makes sense to create a forward to a random internal address.

Yeah, that's fair. Part of me just wants it to be parallel to allowing someone to specify the external network (and have a floating IP created for you). Do you think someone would ever run this first to create a port with the appropriate port forwardings, and then attach it to a node?

In the back of my mind, I was also thinking that the "network designation" logic would be generalized into a function that something like `openstack esi node network list` could use instead of its current heap of optional arguments.

> 
> My only other complaint is that by allowing either an ip address or a (port/network) name, we make option validation much more difficult (we can't reject an invalid ip address, for example, until _after_ we contact the api to see if maybe it was a port name or id).

Yep - that logic definitely becomes more difficult, but I think the alternative is a spaghetti selection of optional arguments which I think looks pretty convoluted to the end user. I think it may be better to err on the side of avoiding that.

> 
> > ```
> > openstack esi port forward create --port <protocol port designation 1>,<protocol port designation 2>,... --from <external network designation> --to <internal network designation>
> > ```
> 
> The thing that bothers me here is that both `--to` and `--from` are required. In the general scheme of things, this suggests they should be positional arguments rather than option arguments...
> 
> ```
> openstack esi port forward create [--port <portspec> [...]] <internal network designation> <external network designation>
> ```
> 
> ...but that loses the contextual clues provided by the option names. Do you have a preference?

Not a strong one - I know plenty of OpenStack CLI commands that actually require optional arguments, but I've always kind of hated that. I'm fine with either choice if the ports are part of the optional argument; that's what really made the command difficult to parse.

> 
> > What do you think?
> 
> I think that overall it sounds like a reasonable plan.

larsks commented:
I've refactored the code in line with our conversation; the following are now all valid commands:

## Create

- Forward a port from an existing floating ip to the same port on an existing internal port:

  ```
  openstack esi port forward create -p 22 10.10.10.10 100.100.100.100
  ```

- Forward a port from a new floating ip to the same port on existing internal port:

  ```
  openstack esi port forward create -p 22 my_internal_port external
  ```

- Forward a port from an existing floating ip to a different port on an internal port:

  ```
  openstack esi port forward create -p 2222:22/tcp 10.10.10.10 100.100.100.100
  ```

- Forward multiple ports:

  ```
  openstack esi port forward create -p 22 -p 80 -p 443 10.10.10.10 100.100.100.100
  ```

## Delete

The `delete` command mirrors the `create` command:

```
openstack esi port forward delete -p 22 10.10.10.10 100.100.100.100
```

## Purge

The `purge` command is unchanged.

tzumainn commented on esiclient/v1/port_forward.py:
```diff
@@ -0,0 +1,422 @@
+                    )
+                internal_network_id = parsed_args.internal_ip_subnet.network_id
+            else:
+                internal_network_id = parsed_args.internal_ip_network.id
```
Reading this section, you don't actually use the network/subnet objects set in `parsed_args.internal_ip_subnet` or `parsed_args.internal_ip_network`; you just grab their UUID. Do the API calls below take a name-or-UUID identifier? If so, would it make sense to avoid the extra API calls to grab the subnet/network, just use the user-provided identifier strings, and allow those API calls to error out if the matching subnet/network can't be found?

larsks replied:
I am a big fan of validating option arguments as early as possible. This avoids situation where you create some resources (say, a floating ip) and then error out afterwards before completing the port forward configuration.

tzumainn replied:
fair enough - I'd suggest revisiting this topic if this command takes a long time to run, but I don't expect this command to be spammed.

tzumainn reviewed (commented):
some initial comments!

larsks commented on esiclient/v1/port_forward.py:
```diff
@@ -0,0 +1,422 @@
+        fip = self.find_or_create_floating_ip(parsed_args)
+        internal_port = self.find_or_create_port(parsed_args)
+
+        if isinstance(
```
No, because based on the data structure a port can have multiple ip addresses. This makes sure we select the appropriate one if an address was specified on the command line (otherwise, we just pick the first one).

tzumainn replied:
ah, gotcha!

larsks commented on esiclient/v1/port_forward.py:
```diff
@@ -0,0 +1,422 @@
+
+
+class NetworkArg:
+    """Handle a command line argumenta that specifies a network name or id"""
```
It sounds like a latin plural.

larsks commented on esiclient/v1/port_forward.py:
```diff
@@ -0,0 +1,422 @@
+                network_id=internal_network_id,
+                fixed_ips=[
+                    {"subnet_id": subnet.id, "ip_address": str(parsed_args.internal_ip)}
+                ],
```
What would you use for the name?

tzumainn replied:
When we attach a network to a node, we use `esi-<node_name>-<network_name>`, but I guess the node isn't available here. Maybe `esi-<network_name>-<modified_ip_address>`, where the colons are replaced by hyphens?

larsks replied:
We don't necessarily know the network name here (if `--internal-ip-subnet` was specified by not `--internal-ip-network`, we only have the network id). Including the ip address in the port name seems like a recipe for confusion if the underlying port is modified to have a different address.

We could set the name to something like `esi-autocreated-<network_id>`, but I'm not really sure that's helpful.

tzumainn replied:
Well, you either have or find the `internal_network_id`, so you can use that to find the network and its name - it is an extra API call though.

I agree that confusion may happen if we name the port after characteristics that are later modified. But I think the benefit outweighs the possible confusion. For example, when attaching a network to a node, putting the node name into the port name makes it far easier to know which ports are for what - and if a user later manually uses that port with a different node, I'd assume that they have the knowledge to modify the port name themselves.

Honestly, I'd also be fine with something like `esi-<network>-<simple date string>` - just anything to make it it a little more distinguishable when someone runs `openstack port list`.

larsks replied:
I just pushed a change that sets it to `esi-autocreated-<address>` for now, despite my misgivings on that front. That at least shows how and why the port was created.

larsks commented on esiclient/v1/port_forward.py:
```diff
@@ -0,0 +1,422 @@
+        )
+        parser.add_argument("--port", "-p", type=PortSpec.from_spec, action="append")
+        parser.add_argument(
+            "internal_ip",
```
I mean, we could call it something like `internal_ip_or_port` or `external_ip_or_network`, I guess, but I think we're becoming guilty of using sentences to name our variables. The name `internal_ip` accurately describes the *purpose* of this argument.

tzumainn replied:
this issue came up for me because I was staring at like 184 for a while wondering how an IP address could be a Neutron port. Could we just use something like `internal_identifier` or `internal_port_identifier`?

larsks replied:
How about:

```diff
diff --git a/esiclient/v1/port_forward.py b/esiclient/v1/port_forward.py
index cf3e58d..6c8d531 100644
--- a/esiclient/v1/port_forward.py
+++ b/esiclient/v1/port_forward.py
@@ -269,12 +269,12 @@ class Create(PortForwardCommand):
         )
         parser.add_argument("--port", "-p", type=PortSpec.from_spec, action="append")
         parser.add_argument(
-            "internal_ip",
+            "internal_ip_descriptor",
             type=AddressOrPortArg(self),
             help="ip address, port name, or port uuid",
         )
         parser.add_argument(
-            "external_ip",
+            "external_ip_descriptor",
             type=AddressOrNetworkArg(self),
             help="ip address or network name",
         )
```

tzumainn replied:
Yeah, that works for me!

larsks commented:
There is now support for `--description` and the beginnings of a test suite.

larsks commented:
The tests are complete. I've squashed everything together. I think I've addressed all the review comments thus far. I was going to add some documentation to this pr, but `doc/commands.md` made me sad and I think I will tackle that in a future pull request.

tzumainn reviewed (commented):
Works pretty well! A few issues:

* when creating port forwardings, `--description` is actually required right now
* if you don't pass in a `--port` argument, you get an unintuitive `'NoneType' object is not iterable` message
* the help text for `port` should maybe state that it can be specified multiple times, and give examples of allowed port formats
* are you sure that an arbitrary protocol name (beyond tcp/udp) shouldn't be allowed? regardless, the error message if you specify something like "foo" isn't the clearest (`error: argument --port/-p: invalid from_spec value: '29:30/foo'`, as opposed to a message that mentions the protocol)
* if you run `openstack esi port forwarding purge` with no arguments, the command "works" with no message - I feel like it may be worth checking this case and asking the user to supply at least one floating IP

larsks commented:
> * when creating port forwardings, `--description` is actually required right now

Ugh, I'm calling that an API foul. Treating `description=None` different from `description` is not specified breaks a pretty common behavior. I've updated the code to handle this situation more gracefully.

> * if you don't pass in a `--port` argument, you get an unintuitive `'NoneType' object is not iterable` message

Yeah, that was ugly. I will fix that.

> * the help text for `port` should maybe state that it can be specified multiple times, and give examples of allowed port formats

I'll update the help text and included a few examples.

> * are you sure that an arbitrary protocol name (beyond tcp/udp) shouldn't be allowed? regardless, the error message if you specify something like "foo" isn't the clearest (`error: argument --port/-p: invalid from_spec value: '29:30/foo'`, as opposed to a message that mentions the protocol)

Sigh. This is `argparse` being annoying. The `ValueError` exception in this case already has a perfectly good error message:

```
>>> Protocol('foo')
[...]
ValueError: 'foo' is not a valid Protocol
>>>
```

But `argparse` is ignoring the exception value. I'll see what the options are for changing how things are handled.

> * if you run `openstack esi port forwarding purge` with no arguments, the command "works" with no message - I feel like it may be worth checking this case and asking the user to supply at least one floating IP

I'll add an error message in this case.

larsks commented:
I think the recent changes resolve all the issues from your last comment.

tzumainn reviewed (commented):
Works great! Can you squash the commits?

larsks commented:
The commits have been squashed.

tzumainn reviewed (approved):
Thanks!
